front/lib/resources/plan_resource.ts

```diff
@@ -0,0 +1,198 @@
+    super(PlanModel, blob);
+  }
+
+  static async makeNew(blob: CreationAttributes<PlanModel>) {
+    const plan = await this.model.create(blob);
+    return new this(this.model, plan.get());
+  }
+
+  static async upsertByPlanCode(blob: CreationAttributes<PlanModel>) {
+    const existing = await this.model.findOne({
+      where: { code: blob.code },
+    });
+
+    if (existing) {
+      await existing.update(blob);
+      return new Ok(new PlanResource(this.model, existing.get()));
+    }
+    const plan = await PlanResource.makeNew(blob);
+    return new Ok(plan);
+  }
+
+  static async fetchByPlanCode(planCode: string): Promise<PlanResource | null> {
+    const plan = await this.model.findOne({
+      where: {
+        code: planCode,
+      },
+    });
+    return plan ? new this(this.model, plan.get()) : null;
+  }
+
+  // fetch all plans associated to a given set of subscriptions
+  static async fetchBySubscriptionModels(
+    subscriptions: Subscription[]
+  ): Promise<PlanResource[] | null> {
+    const plans = await this.model.findAll({
+      where: {
+        id: subscriptions.map((s) => s.planId),
+      },
+    });
+    return plans.map((plan) => new this(this.model, plan.get()));
+  }
+
+  static async fetchAll(
+    auth: Authenticator,
+    { limit, order }: ResourceFindOptions<PlanModel> = {}
+  ): Promise<PlanResource[]> {
+    if (!auth.isDustSuperUser()) {
+      throw new Error("Cannot fetch all plans : not allowed.");
+    }
+
+    const plans = await this.model.findAll({
+      limit,
+      order,
+    });
+    return plans.map((plan) => new this(this.model, plan.get()));
+  }
+
+  static async fetchFirst(): Promise<PlanResource | null> {
+    const plan = await this.model.findOne({});
+    return plan ? new this(this.model, plan.get()) : null;
+  }
+
+  async delete(
+    auth: Authenticator,
+    { transaction }: { transaction?: Transaction } = {}
+  ): Promise<Result<undefined, Error>> {
+    await this.model.destroy({
+      where: {
+        id: this.id,
+      },
+      transaction,
+    });
+    return new Ok(undefined);
+  }
+
+  // This function is used to set the message limits for a given plan.
+  static async setMessageLimitsForPlan(
+    data: Pick<Attributes<PlanModel>, "maxMessages" | "maxMessagesTimeframe">,
+    planCode: PlanModel["code"]
+  ): Promise<Result<undefined, Error>> {
+    try {
+      await this.model.update(data, {
+        where: { code: planCode },
+      });
+      return new Ok(undefined);
+    } catch (err) {
+      return new Err(err as Error);
+    }
+  }
+
+  // This function is used to reset the plan with new data.
+  async updateByPlanCode(
+    planCode: string,
+    planData: Omit<Attributes<PlanModel>, "id" | "createdAt" | "updatedAt">
+  ): Promise<Result<undefined, Error>> {
+    await this.model.update(planData, {
+      where: {
+        code: planCode,
+      },
+    });
+    return new Ok(undefined);
+  }
```

Comment:
Does not make sense to not use the current resource id on a class's method. Suggesting something like:
```suggestion
  async updateByPlanCode(
    planCode: string,
    planData: Omit<Attributes<PlanModel>, "id" | "createdAt" | "updatedAt">
  ): Promise<[affectedCount: number]> {
    return  this.update(planData);
   }
```